sycl/test-e2e/lit.cfg.py

```diff
@@ -413,17 +411,28 @@
     else:
         config.substitutions.append(('%xptifw_lib', " -L{} -lxptifw -I{} ".format(xptifw_lib_dir, xptifw_includes)))
 
-
-llvm_tools = ["llvm-spirv", "llvm-link"]
-for llvm_tool in llvm_tools:
-  llvm_tool_path = find_executable(llvm_tool)
-  if llvm_tool_path:
-    lit_config.note("Found " + llvm_tool)
-    config.available_features.add(llvm_tool)
-    config.substitutions.append( ('%' + llvm_tool.replace('-', '_'),
-                                  os.path.realpath(llvm_tool_path)) )
-  else:
-    lit_config.warning("Can't find " + llvm_tool)
+# Substitute standard lit tools: FileCheck, not, etc. Do this down here as we
+# need the FileCheck substitution to be ordered after the substitutions using
```

Comment:
Not a lit expert, but I think I've seen sorting of substitutes inside its implementation. Might it be something else going on here that requires that order? Not a blocker for this review.

Reply:
So it turns out it's not about sorting, but about the recursive nature of substitutions.

When `config.recursiveExpansionLimit` is not set, the substitution process runs over the list of substitutions and applies each. When it's set, it'll perform this process iteratively. We need to set the recursive limit to ensure that the `FileCheck` used in the CHECK placeholders is substituted.

This is why I found that I had to call `llvm_config.use_default_substitutions` so far down in the file because without recursive substitution I had to have the `FileCheck` substitution placed after the CHECK placeholders in the list.

So I think we should unconditionally set `config.recursiveExpansionLimit = 10`, not just when `config.run_launcher` is set.

We don't even really want to call `llvm_config.use_default_substitutions` because it treats its tool substitutions as fatal, which we don't want. I do think it's helpful to have these tools substituted, as then the RUN lines as printed can be accurately reproduced by users: there's a difference when lit runs `| FileCheck` and when users run `| FileCheck`, due to lit's implicit PATH and other environment variables.